src/mig/modules/timedrift/timedrift.go

```diff
@@ -0,0 +1,330 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Contributor: Julien Vehent jvehent@mozilla.com [:ulfr]
+package timedrift
+
+import (
+	"bufio"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"mig"
+	"net"
+	"os"
+	"time"
+)
+
+// init is called by the Go runtime at startup. We use this function to
+// register the module in a global array of available modules, so the
+// agent knows we exist
+func init() {
+	mig.RegisterModule("timedrift", func() interface{} {
+		return new(Runner)
+	})
+}
+
+// Runner gives access to the exported functions and structs of the module
+type Runner struct {
+	Parameters params
+	Results    results
+}
+
+type results struct {
+	FoundAnything bool        `json:"foundanything"`
+	Success       bool        `json:"success"`
+	Elements      checkedtime `json:"elements"`
+	Statistics    statistics  `json:"statistics"`
+	Errors        []string    `json:"errors"`
+}
+
+// a simple parameters structure, the format is arbitrary
+type params struct {
+	Drift string `json:"drift"`
+}
+
+type checkedtime struct {
+	HasCheckedDrift bool     `json:"hascheckeddrift"`
+	IsWithinDrift   bool     `json:"iswithindrift,omitempty"`
+	Drifts          []string `json:"drifts,omitempty"`
+	LocalTime       string   `json:"localtime"`
+}
+
+type statistics struct {
+	ExecTime string     `json:"exectime"`
+	NtpStats []ntpstats `json:"ntpstats,omitempty"`
+}
+
+type ntpstats struct {
+	Host      string    `json:"host"`
+	Time      time.Time `json:"time"`
+	Latency   string    `json:"latency"`
+	Drift     string    `json:"drift"`
+	Reachable bool      `json:"reachable"`
+}
+
+func (r Runner) ValidateParameters() (err error) {
+	if r.Parameters.Drift != "" {
+		_, err = time.ParseDuration(r.Parameters.Drift)
+	}
+	return err
+}
+
+func (r Runner) Run(Args []byte) string {
+	var (
+		stats statistics
+		ct    checkedtime
+		drift time.Duration
+	)
+	ct.LocalTime = time.Now().Format(time.RFC3339Nano)
+	t1 := time.Now()
+	err := json.Unmarshal(Args, &r.Parameters)
+	if err != nil {
+		r.Results.Errors = append(r.Results.Errors, fmt.Sprintf("%v", err))
+		return r.buildResults()
+	}
+	err = r.ValidateParameters()
+	if err != nil {
+		r.Results.Errors = append(r.Results.Errors, fmt.Sprintf("%v", err))
+		return r.buildResults()
+	}
+	// if drift is not set, skip the ntp test
+	if r.Parameters.Drift == "" {
+		r.Results.FoundAnything = true
+		goto done
+	}
+	drift, err = time.ParseDuration(r.Parameters.Drift)
+	if err != nil {
+		r.Results.Errors = append(r.Results.Errors, fmt.Sprintf("%v", err))
+		return r.buildResults()
+	}
+	// assume host has synched time and set to false if not true
+	ct.IsWithinDrift = true
+	for _, host := range NtpPool {
+		t, lat, err := GetTimeFromHost(host)
```

Comment:
I thought about it for the same reason, but decided that it's not worth dealing with parallelism. Most of the time, NTP servers will respond in ~30ms, so total exec time is around 120ms, and max time is 4*5s=20s for the rare case where the host doesn't have UDP connectivity to the internet, or NTP servers are down.